src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,514 @@ impl Display for SnapKey {
 }
 
-impl SnapFile {
-    fn new<T: Into<PathBuf>>(snap_dir: T,
-                             size_track: Arc<RwLock<u64>>,
-                             is_sending: bool,
-                             key: &SnapKey)
-                             -> io::Result<SnapFile> {
-        let mut file_path = snap_dir.into();
-        if !file_path.exists() {
-            try!(fs::create_dir_all(file_path.as_path()));
-        }
-        let prefix = if is_sending {
-            SNAP_GEN_PREFIX
-        } else {
-            SNAP_REV_PREFIX
-        };
-        let file_name = format!("{}_{}.snap", prefix, key);
-        file_path.push(&file_name);
-
-        let mut f = SnapFile {
-            file: file_path,
-            digest: Digest::new(crc32::IEEE),
-            size_track: size_track,
-            tmp_file: None,
-        };
-        try!(f.init());
-        Ok(f)
-    }
-
-    pub fn init(&mut self) -> io::Result<()> {
-        if self.exists() || self.tmp_file.is_some() {
-            return Ok(());
+// A helper function to copy snapshot.
+// Only used in tests.
+pub fn copy_snapshot(mut from: Box<Snapshot>, mut to: Box<Snapshot>) -> io::Result<()> {
+    if !to.exists() {
+        try!(io::copy(&mut from, &mut to));
+        try!(to.save());
+    }
+    Ok(())
+}
+
+fn need_to_pack(cf: &str) -> bool {
+    // Data in CF_RAFT should be excluded for a snapshot.
+    // Check CF_RAFT here and skip it, even though CF_RAFT should not occur
+    // in the range [begin_key, end_key) of a RocksDB snapshot usually.
+    cf != CF_RAFT
+}
+
+mod v1 {
+    use std::cmp;
+    use std::io::{self, Read, Write, ErrorKind};
+    use std::fs::{self, File, OpenOptions, Metadata};
+    use std::path::{Path, PathBuf};
+    use std::sync::{Arc, RwLock};
+    use std::str;
+    use std::time::Instant;
+    use crc::crc32::{self, Digest, Hasher32};
+    use byteorder::{BigEndian, WriteBytesExt, ReadBytesExt};
+    use rocksdb::{Writable, WriteBatch};
+    use kvproto::metapb::Region;
+    use kvproto::raft_serverpb::RaftSnapshotData;
+    use raft;
+    use util::{HandyRwLock, rocksdb};
+    use util::codec::bytes::{BytesEncoder, CompactBytesDecoder};
+
+    use super::super::engine::{Snapshot as DbSnapshot, Iterable};
+    use super::super::keys::{self, enc_start_key, enc_end_key};
+    use super::super::util;
+    use super::{SNAP_GEN_PREFIX, SNAP_REV_PREFIX, TMP_FILE_SUFFIX, Result, SnapKey, Snapshot,
+                ApplyContext, check_abort, need_to_pack};
+
+    pub const CRC32_BYTES_COUNT: usize = 4;
+    const DEFAULT_READ_BUFFER_SIZE: usize = 4096;
+
+    /// A structure represents the snapshot.
+    ///
+    /// All changes to the file will be written to `tmp_file` first, and use
+    /// `save` method to make them persistent. When saving a crc32 checksum
+    /// will be appended to the file end automatically.
+    pub struct Snap {
+        file: PathBuf,
+        digest: Option<Digest>,
```

Comment:
I think just `Digest` is ok. Using `Option` here just make the struct more larger.